<div style="background-color:LightBlue; text-align:center; padding:20px;">
    <h2 style="color:black; font-family: Verdana, sans-serif;"><strong>Multi-Agent Reinforcement Learning Project - Atari</strong></h2>
    <p style="font-size: 14px; color: black; font-family: Verdana, sans-serif;">
        <table style="margin: auto; border-collapse: collapse; color: black;">
            <tr>
                <th style="border: 0;">Names</th>
                <th style="border: 0;">GitHub Username</th>
            </tr>
            <tr>
                <td style="border: 0;">Rogier Gernaat</td>
                <td style="border: 0;">RogierHHS</td>
            </tr>
            <tr>
                <td style="border: 0;">Daan Eising</td>
                <td style="border: 0;">DaanEising</td>
            </tr>
            <tr>
                <td style="border: 0;">Julia Boschman</td>
                <td style="border: 0;">JuliaBoschman</td>
            </tr>
            <tr>
                <td style="border: 0;">Jort Akershoek</td>
                <td style="border: 0;">JortAkershoek</td>
            </tr>
        </table>
    </p></div>

<div style="display: flex; justify-content: center; align-items: center; margin-top: 10px;">
    <img src="" alt="fotoj van onze opdracht/GIF" style="width: 1000px; height: auto;">
</div>


- ***Docent***: Vikram Radhakrishnan
- ***Datum***: 08-04-2025


---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong> Inhoudsopgave (Prototype morgen kunnen we hem nog ff aanpassen of anders indelen) </strong></h2>
</div>

## **Inhoudsopgave**

1. [H1: Inleiding](#1.0)
   - [&sect;1.1: Imports en Setup](#1.1)  
  
2. [H2: Kiezen van Algoritme](#2.0)  
   - [&sect;2.1: Kiezen van RL-Algoritme](#2.1)    
   - [&sect;2.3: Kiezen van Trainingsstrategie](#2.3)  

3. [H3: Probleemdefinitie](#3.0)  
   - [&sect;3.1: Wat is het probleem?](#3.1)   

4. [H4: Ontwerp en Implementatie](#4.0)  
   - [&sect;4.1: Baseline strategie ontwikkelen](#4.1)  
   - [&sect;4.2: Selectie van DRL algoritme en frameworks](#4.2)   
   - [&sect;4.3: Implementatie MARL-agent](#4.3)  

5. [H5: Training en Hyperparameter Search](#5.0)  
   - [&sect;5.1: Training](#5.1)  
   - [&sect;5.2: Selectie en tuning van hyperparameters](#5.2)  

6. [H6: Evaluatie en Vergelijking](#6.0)  
   - [&sect;6.1: Evaluatie t.o.v. baseline](#6.1)  
   - [&sect;6.2: Analyse met metrics](#6.2)  
   - [&sect;6.3: Visualisatie van resultaten](#6.3)  

7. [H7: Rapportage en Reflectie](#7.0)  
   - [&sect;7.1: Methodologie en aanpak](#7.1)  
   - [&sect;7.2: Samenvatting van resultaten](#7.2)  
   - [&sect;7.3: Reflectie op model, prestaties en uitbreidingsmogelijkheden](#7.3)  

8. [H8: Literatuurlijst](#8.0)  

9. [H9: Beoordelingscriteria](#9.0)



---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H1: Inleiding </strong></h2>
</div>

## inleiding bladiebladiebla

<a name='1.1'></a>
<h3>&sect;1.1: Imports en Setup</h3>

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Pip install's</strong>
</div>



In [ ]:
# Install the necessary libraries

!pip install gym
!pip install stable_baselines3
!pip install gymnasium[atari]
!pip install pettingzoo[atari]
!pip install "autorom[accept-rom-license]"
!pip install --find-links dist/ --no-cache-dir AutoROM[accept-rom-license]
!pip install supersuit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.5/184.5 kB 8.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 111.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 89.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 49.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 96.2 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstall

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Importeren van de library's</strong>
</div>



In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import sys
import gymnasium as gym
from stable_baselines3 import PPO
from google.colab import drive
# Start AutoROM

!AutoROM

# Import libraries
from pettingzoo.atari import warlords_v3
from pettingzoo.utils import BaseParallelWrapper
from collections import defaultdict, Counter
import importlib
import os
import imageio

AutoROM will download the Atari 2600 ROMs.
They will be installed to:
	/usr/local/lib/python3.11/dist-packages/AutoROM/roms
	/usr/local/lib/python3.11/dist-packages/multi_agent_ale_py/roms

Existing ROMs will be overwritten.

I own a license to these Atari 2600 ROMs.
I agree to not distribute these ROMs and wish to proceed: [Y/n]: y
Installed /usr/local/lib/python3.11/dist-packages/AutoROM/roms/adventure.bin
Installed /usr/local/lib/python3.11/dist-packages/multi_agent_ale_py/roms/adventure.bin
Installed /usr/local/lib/python3.11/dist-packages/AutoROM/roms/air_raid.bin
Installed /usr/local/lib/python3.11/dist-packages/multi_agent_ale_py/roms/air_raid.bin
Installed /usr/local/lib/python3.11/dist-packages/AutoROM/roms/alien.bin
Installed /usr/local/lib/python3.11/dist-packages/multi_agent_ale_py/roms/alien.bin
Installed /usr/local/lib/python3.11/dist-packages/AutoROM/roms/amidar.bin
Installed /usr/local/lib/python3.11/dist-packages/multi_agent_ale_py/roms/amidar.bin
Installed /usr/local/

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Mounten van colab drive en inladen van de agents</strong>
</div>

In [ ]:
drive.mount('/content/drive')
sys.path.append('/content/drive/MyDrive/Autonomous Systems')
print(os.listdir('/content/drive/MyDrive/Autonomous Systems'))

Mounted at /content/drive



---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H2: Kiezen van Algoritme </strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;2.1: Kiezen van RL-Algoritme</h3>

a.	Jouw agent zal deelnemen en interacten met andere agenten in de Atari-omgeving "Warlords".

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Agent 1: (Rogier) DQN - Deep Q-Network</strong>
</div>



## §2.1: Kiezen van RL-Algoritme

In deze opdracht staat de Atari-game **Warlords** centraal: een klassieke arcadegame waarbij vier spelers (agents) gelijktijdig strijden op een speelveld. Iedere agent verdedigt zijn eigen kasteel en probeert de anderen uit te schakelen, wat resulteert in een typisch **multi-agent scenario** met zowel competitie als wisselende interactiepatronen tussen agents. Voor deze complexe en dynamische omgeving is het belangrijk om een reinforcement learning-algoritme te kiezen dat bewezen effectief is bij problemen met een hoge mate van onzekerheid, veel mogelijke toestanden en meerdere spelers.

We kiezen voor het **Proximal Policy Optimization (PPO)** algoritme als basis voor onze agent. PPO is een krachtig, modern en veelgebruikt algoritme binnen deep reinforcement learning. Het is ontworpen voor stabiliteit en efficiëntie bij het optimaliseren van beleid (policies), en heeft uitstekende prestaties laten zien in visuele, dynamische omgevingen zoals Atari-games. PPO leert direct van ruwe pixeldata via een convolutioneel neuraal netwerk en maakt gebruik van policy-gradient updates die gecontroleerd worden uitgevoerd om instabiliteit te voorkomen.

### Waarom PPO?

- **Geschikt voor visuele input:**  
  PPO werkt uitstekend met ruwe spelbeelden (frames) als input, en kan daardoor zelfstandig complexe strategieën ontwikkelen zonder handmatige feature engineering. De convolutionele neurale netwerken van PPO zijn uitermate geschikt voor het herkennen van visuele patronen in Atari-omgevingen.

- **Stabiel en robuust leren:**  
  Door het gebruik van trust-region updates (clipping), mini-batch learning en advantage schatting (GAE), blijft het leerproces gecontroleerd en raakt het model minder snel verstrikt in abrupte of onstabiele policy-wijzigingen. Dit is cruciaal in chaotische multi-agent omgevingen zoals Warlords.

- **Multi-agent compatibiliteit:**  
  PPO laat zich eenvoudig toepassen op multi-agent settings, bijvoorbeeld via parameter sharing (één policy voor meerdere agents) of door individuele policies te trainen per agent. Dit maakt het flexibel inzetbaar voor uiteenlopende experimenten.

- **Breed onderzocht en veel gebruikt:**  
  PPO behoort tot de standaardbenchmarks in reinforcement learning en is in tal van studies succesvol ingezet voor Atari-omgevingen en multi-agent settings. Er zijn veel goed onderhouden frameworks beschikbaar (zoals Stable-Baselines3 en PettingZoo), wat snelle en correcte implementatie mogelijk maakt.

### Concreet voordeel voor deze opdracht

- **Algoritmische robuustheid:**  
  PPO is bijzonder effectief in dynamische en onvoorspelbare omgevingen waar het gedrag van andere agents voortdurend verandert. Dankzij gecontroleerde policy-updates en efficiënte verwerking van ervaringen is PPO in staat om robuuste strategieën te ontwikkelen.

- **Vergelijkbaarheid:**  
  PPO is een veelgebruikte standaard in de literatuur rondom Atari en multi-agent reinforcement learning. Hierdoor zijn onze resultaten goed te vergelijken met bestaande benchmarks en alternatieve algoritmes (zoals DQN of random policies).

- **Transparantie en reproduceerbaarheid:**  
  Dankzij de brede adoptie en solide implementaties in frameworks als Stable-Baselines3 en PettingZoo zijn onze experimenten eenvoudig te reproduceren, uit te breiden en te valideren door andere studenten of onderzoekers.


<div style="background-color:white; color:black; padding: 10px;">
    <strong>Agent 2: (Julia) PPO - Proximal Policy Optimization</strong>
</div>



### **Uitleg van PPO**

Proximal Policy Optimization (PPO) is een geavanceerd reinforcement learning-algoritme dat veel wordt gebruikt voor het trainen van agents in complexe omgevingen zoals Atari-spellen. PPO verbetert eerdere policy-gradient-methodes door bij elke stap de aanpassing van het beleid (“policy”) te beperken. Hierdoor wordt het leerproces stabieler en is de kans kleiner dat de agent ineens “vergeet” wat hij geleerd heeft.

PPO werkt door het beleid steeds een klein beetje aan te passen op basis van ervaringen uit de omgeving. Hierdoor leert de agent efficiënter en zijn de resultaten vaak beter reproduceerbaar. PPO is ook geschikt voor situaties met hoge-dimensionale input, zoals beelden, en werkt goed in multi-agent omgevingen zoals Warlords.

DhanushKumar (2024)

### **Motivatie**

Voor deze opdracht heb ik gekozen voor het algoritme Proximal Policy Optimization (PPO). De belangrijkste reden hiervoor is dat PPO bekend staat om zijn stabiliteit en efficiëntie bij het trainen van agents in complexe omgevingen met hoge-dimensionale input, zoals de Atari-game Warlords, Schulman et al. (2017). Omdat PPO de aanpassingen aan het beleid per stap beperkt, blijft het leerproces gecontroleerd en voorkom je dat de agent tijdens het trainen “vergeet” wat eerder geleerd is. Dit is vooral belangrijk in multi-agent omgevingen, waar het gedrag van andere agents de situatie voortdurend beïnvloedt.

Daarnaast is PPO eenvoudig te implementeren dankzij bestaande libraries zoals stable-baselines3, waardoor het mogelijk is om snel te experimenteren met verschillende hyperparameters. De standaardwaarden die ik voor de belangrijkste hyperparameters heb gekozen, zijn gebaseerd op aanbevelingen uit de literatuur en eerdere succesvolle toepassingen in vergelijkbare omgevingen (The 37 Implementation Details Of Proximal Policy Optimization · The ICLR Blog Track, 2022). PPO is bovendien goed schaalbaar, waardoor het geschikt is om in een multi-agent setting zoals Warlords verschillende agents onafhankelijk van elkaar te trainen en te vergelijken.

Door deze eigenschappen is PPO naar mijn mening de meest geschikte keuze voor deze opdracht, omdat het zorgt voor betrouwbare leerresultaten en flexibiliteit biedt bij het uitvoeren van experimenten met verschillende agents en trainingsinstellingen.

### **Aanpak**

Voor deze opdracht heb ik een PPO-agent geïmplementeerd met behulp van de stable-baselines3 library. PPO is gekozen vanwege de stabiele prestaties en de robuustheid bij het trainen in omgevingen met beeldinput, zoals “Warlords”. De agent gebruikt een convolutioneel neuraal netwerk om de observaties te verwerken.
De belangrijkste hyperparameters zijn gekozen op basis van aanbevolen waarden uit wetenschappelijke literatuur voor Atari-omgevingen, maar kunnen verder geoptimaliseerd worden.
Mijn implementatie maakt het makkelijk om het model op te slaan, opnieuw te laden en te evalueren, zodat experimenten goed reproduceerbaar zijn.

<a name='1.1'></a>
<h3>&sect;2.3: Kiezen van Trainingsstrategie</h3>

b.	Kies een geschikt RL-algoritme en trainingsstrategie voor je agent. Beschrijf en motiveer je keuzes.

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Agent 1: (Rogier) DQN - Deep Q-Network </strong>
</div>



### Trainingsstrategie in Multi-Agent Warlords

Om optimaal gebruik te maken van het gekozen algoritme, hanteren we een trainingsstrategie die inspeelt op de uitdagingen van multi-agent reinforcement learning:

1. **Multi-agent omgeving:**  
   In Warlords zijn altijd meerdere agents actief. Onze PPO-agent wordt getraind door herhaaldelijk games te spelen tegen vooraf ingestelde tegenstanders, zoals random agents of (indien gewenst) andere PPO-agents. Dit zorgt ervoor dat de agent leert in een realistische, competitieve setting.

2. **Variatie in tegenstanders:**  
   Door de agent bloot te stellen aan verschillende typen tegenstanders (van random tot geavanceerd), voorkomen we dat de agent eenzijdige strategieën aanleert. Zo ontwikkelt de agent robuustere, meer generaliseerbare strategieën.

3. **Observatie- en actie-preprocessing:**  
   We gebruiken beeldverwerkingstechnieken zoals grijswaardenconversie, rescaling en stacking van frames. Dit zorgt ervoor dat de agent alleen de essentiële informatie uit het spel verwerkt, waardoor het leerproces efficiënter verloopt.

4. **On-policy leren & rollouts:**  
   PPO werkt met zogenoemde rollouts: de agent verzamelt verse trajecten door de omgeving en leert direct van deze actuele ervaringen. Dit betekent dat elke policy-update gebaseerd is op recent gedrag, wat bijdraagt aan stabiliteit en effectieve policy-veranderingen.

5. **Regelmatige evaluatie en hyperparameter tuning:**  
   Tijdens de training evalueren we regelmatig de prestaties van de agent door deze te laten spelen tegen de baseline (zoals een random agent). Daarnaast experimenteren we met verschillende hyperparameters, zoals learning rate, batch size, en de clipping-range van PPO, om de optimale instellingen te bepalen.

6. **Logging en visualisatie:**  
   Alle trainingsresultaten (zoals reward curves, winpercentages, etc.) worden gelogd en gevisualiseerd. Hierdoor kunnen we het leerproces volgen en analyseren, en waar nodig de strategie bijstellen.

### Samenvattend

Deze aanpak zorgt ervoor dat onze PPO-agent niet alleen leert van het eigen gedrag, maar zich ook kan aanpassen aan verschillende soorten tegenstanders. Door systematisch te trainen, evalueren en hyperparameters te tunen, halen we het maximale uit ons algoritme en maken we de voordelen van reinforcement learning in een multi-agent context helder zichtbaar.


<div style="background-color:white; color:black; padding: 10px;">
    <strong>Agent 2: (Julia) PPO - Proximal Policy Optimization (Hier mist denk ik nog wat of ik heb het niet goed doorgelezen)(Julia morgen)</strong>
</div>




---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H3: Probleemdefinitie </strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;3.1: Wat is het probleem?</h3>

c.	Resultaat: Een helder uitgewerkte motivatie in je rapport (“Inleiding & Probleemanalyse”).

#### Probleemanalyse
De opkomst van multi-agent omgevingen in toepassingen zoals robotica, games en logistiek vraagt om slimme algoritmes die kunnen concurreren én samenwerken. In de praktijk betekent dit dat agents hun strategieën continu moeten bijstellen op basis van het gedrag van andere agents in hun omgeving. In de Atari-game Warlords komen vier agents tegelijkertijd in actie, waarbij hun succes afhankelijk is van zowel hun eigen keuzes als die van hun tegenstanders.

Single-agent reinforcement learning is onvoldoende, omdat hierbij wordt aangenomen dat de omgeving stationair is (niet verandert door anderen). In multi-agent settings verandert de omgeving echter continu, omdat andere agents ook leren en hun gedrag aanpassen. Dit vraagt om een benadering waarbij agents niet alleen leren van hun eigen ervaringen, maar ook van de interacties met anderen.

Met multi-agent reinforcement learning (MARL) kunnen agents hun beleid optimaliseren terwijl ze rekening houden met de strategieën van anderen. Hierdoor ontstaan vaak complexe en onverwachte gedragingen die in single-agent settings niet mogelijk zijn. Bovendien kunnen MARL-methoden worden ingezet om situaties te modelleren waarin competitie, samenwerking of beide tegelijk nodig zijn.

##### Relevantie van het probleem
- In veel echte omgevingen zijn meerdere autonome beslissers actief (bijvoorbeeld zelfrijdende auto's in verkeer).
- Het ontwerpen van robuuste agents in zulke settings helpt bij het ontwikkelen van realistische, schaalbare en adaptieve AI-systemen.
- In de context van games als Warlords kan MARL inzichten bieden in hoe intelligente strategieën en tegenstrategieën ontstaan in competitieve settings.

##### Samenvatting probleemstelling (één zin):
"Hoe kunnen we effectieve, lerende agents ontwikkelen die optimaal presteren in een competitieve multi-agent omgeving, waarbij rekening wordt gehouden met de voortdurende interactie en dynamiek tussen verschillende agents?"

Scharwächter (2024)

#### Doelstelling
Het doel van deze opdracht is om een multi-agent reinforcement learning systeem te ontwerpen, implementeren en evalueren voor de Atari Warlords-omgeving. Dit gebeurt door vier verschillende MARL-algoritmes (PPO, MADDPG, [keuze Peet] en [keuze Jort]) te trainen en hun prestaties te vergelijken met een baseline. Het eindresultaat is een reproduceerbaar systeem en een rapport met een diepgaande analyse van de werking en effectiviteit van de gekozen algoritmen.


---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H4: Ontwerp en Implementatie</strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;4.1: Baseline strategie ontwikkelen</h3>

a.	Ontwikkel een baseline zoals bijvoorbeeld een rule-based policy een random policy of een andere simpele heuristiek.  

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Baseline: Random Agent</strong>
</div>

Als startpunt voor het vergelijken van verschillende algoritmes is het belangrijk om een baseline te definiëren. In deze opdracht gebruiken we een random agent als baseline. Dit is een agent die bij elke stap willekeurig een van de mogelijke acties kiest, ongeacht de observatie of situatie in het spel.

**Waarom een random agent als baseline?**

Een random agent biedt een objectief referentiepunt: het laat zien wat de prestaties zouden zijn zonder enige vorm van intelligentie, strategie of leren. Door de resultaten van geavanceerdere agents (zoals een rule-based agent of een reinforcement learning agent zoals PPO) te vergelijken met deze random agent, kun je duidelijk aantonen of jouw aanpak daadwerkelijk beter presteert dan toeval.

**Implementatie**

De implementatie van de random agent is heel eenvoudig. De agent kiest telkens een willekeurige actie uit het totale aantal toegestane acties van de omgeving. In het geval van Atari Warlords zijn dit bijvoorbeeld zes mogelijke acties.


In [ ]:
class AgentRandomPolicy:
    def act(self, observation):
        # Return a random action (6 possible in ALE Warlords)
        return np.random.randint(6)


<a name='1.1'></a>
<h3>&sect;4.2: Selectie van DRL algoritme en frameworks</h3>

b.	Kies een passend DRL-algoritme, die geschikt is voor een multi-agent setting. Kies geschikte packages en frameworks.

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Dit is uitgewerkt maar niet iedereen zijn agent staat er nu bij sinds ik dat nog niet weet en of je packages etc zijn uitgelegd. Iedereen moet dat hier zelf in toevoegen <strong> </div>



### Keuze van Algoritme

Voor deze opdracht, waarin de agent moet presteren in de **multi-agent omgeving van Atari Warlords**, kiezen we voor het **Proximal Policy Optimization (PPO)** algoritme. PPO is momenteel één van de meest gebruikte en robuuste algoritmes voor deep reinforcement learning, vooral geschikt voor problemen met complexe visuele input en multi-agent interactie. De belangrijkste voordelen van PPO:

- **Stabiel leren van ruwe pixels:** PPO maakt gebruik van convolutionele neurale netwerken (CNN’s) om direct van visuele observaties te leren, zonder handmatig feature engineering.
- **Uitstekende prestaties in Atari-omgevingen:** PPO heeft zich bewezen als benchmark-algoritme in veel Atari-games, mede dankzij de balans tussen exploratie en exploitatie.
- **Direct geschikt voor multi-agent settings:** PPO kan eenvoudig worden ingezet met *parameter sharing* (één policy voor meerdere agents) of individuele policies per agent, wat het flexibel maakt voor uiteenlopende MARL-experimenten.

### Keuze & Motivatie

In deze opdracht trainen we een **PPO-agent** in de multi-agent Warlords-omgeving. Hierbij nemen onze agenten het op tegen ingebouwde tegenstanders, zoals random agents. Zo kunnen we de kracht van deep RL aantonen in vergelijking met simpele baseline strategieën.

#### Waarom PPO in deze context?

- **Stabiele policy learning:** PPO minimaliseert het risico op instabiliteit door gecontroleerde policy-updates (clipping), wat vooral belangrijk is in chaotische multi-agent omgevingen.
- **Visual input:** Warlords levert observaties als pixeldata, wat naadloos aansluit op de CNN-architectuur van PPO.
- **Flexibiliteit:** PPO werkt zowel met discrete als continue actie-ruimtes en laat zich makkelijk combineren met moderne MARL frameworks.

---

### Packages en Frameworks

Voor de implementatie en evaluatie maken we gebruik van de volgende frameworks en libraries:

#### 1. **Stable-Baselines3**
- Biedt een krachtige en stabiele implementatie van PPO, met uitgebreide ondersteuning voor logging en evaluatie.
- Direct compatibel met vectorized en custom omgevingen.

#### 2. **PettingZoo**
- De standaard voor multi-agent reinforcement learning omgevingen, met een ruime keuze aan benchmarkomgevingen zoals Atari Warlords.
- Zorgt voor een uniforme interface en maakt snelle experimentatie mogelijk.

#### 3. **Supersuit**
- Bibliotheek voor preprocessing (o.a. frames resizen, kleuren reduceren en frame stacking), essentieel voor efficiënte training op visuele data.

---

### Samenvatting van de strategie

- We trainen een PPO-agent in de **PettingZoo Atari Warlords** omgeving tegen baseline agents (zoals random).
- PPO stelt ons in staat om efficiënt en stabiel te leren in deze multi-agent setting, met als doel beter te presteren dan eenvoudige baselines.
- Door te bouwen op bewezen frameworks en preprocessing pipelines is onze aanpak **reproduceerbaar, schaalbaar en robuust**.

---


<a name='1.1'></a>
<h3>&sect;4.3: Implementatie MARL-agent</h3>

d.	Resultaat: Een werkend MARL-systeem dat klaar is voor training en evaluatie.

<div style="background-color:white; color:black; padding: 10px;">
    <strong>(Rogier) PPO Agent ( DEZE FUNCTIE NOG ZETTEN BIJ DE TRAINING STAP )</strong>
</div>

In de onderstaande cel definiëren we de klasse MARLAgentPPO, een modulair systeem dat geschikt is voor het trainen van PPO-agenten in een multi-agent setting met behulp van de PettingZoo-omgeving warlords_v3. De klasse bevat methoden voor het instellen van de omgeving, het trainen van het model, en het opslaan of laden van een getraind PPO-model.

 -   De omgeving wordt gepreprocessed met behulp van Supersuit-wrappers (zoals black_death, color_reduction, resizing en frame stacking).

-  Vervolgens wordt de omgeving geconverteerd naar een vectorized environment die compatibel is met Stable-Baselines3.

-   De train()-methode traint het PPO-model met een CnnPolicy, geschikt voor de beeldinvoer van Atari-games.

-  Met save() en load() kunnen getrainde modellen eenvoudig worden opgeslagen of ingeladen.

Deze klasse maakt het mogelijk om met ons MARL-systeem op een reproduceerbare en schaalbare manier PPO-agenten te trainen in een multi-agent omgeving.

(PPO — Stable Baselines3 2.7.0a0 Documentation, z.d.)

# MARL PPO met meer parameters

In [ ]:
import time
import supersuit as ss
from stable_baselines3 import PPO
from stable_baselines3.ppo import CnnPolicy
from pettingzoo.atari import warlords_v3

class MARLAgentPPO:
    """
    Multi-Agent RL systeem voor de Atari Warlords omgeving.
    Ondersteunt setup, training, evaluatie en het opslaan/laden van PPO-modellen.
    """

    def __init__(
        self,
        num_vec_envs=2,
        frame_stack=4,
        x_size=84,
        y_size=84,
        batch_size=256,
        total_timesteps=4_000_000,
        verbose=1
    ):
        """
        Initialiseer de omgeving en hyperparameters.
        """
        self.num_vec_envs = num_vec_envs
        self.frame_stack = frame_stack
        self.x_size = x_size
        self.y_size = y_size
        self.batch_size = batch_size
        self.total_timesteps = total_timesteps
        self.verbose = verbose

        self.env = self._make_env()
        self.vec_env = self._make_vec_env()
        self.model = None

    def _make_env(self):
        """
        Zet de Warlords-omgeving klaar met preprocessing-wrappers.
        """
        env = warlords_v3.parallel_env()
        env = ss.black_death_v3(env)
        env = ss.color_reduction_v0(env, mode='full')
        env = ss.resize_v1(env, x_size=self.x_size, y_size=self.y_size)
        env = ss.frame_stack_v1(env, self.frame_stack)
        return env

    def _make_vec_env(self):
        """
        Converteer naar een vectorized environment voor Stable-Baselines3.
        """
        vec_env = ss.pettingzoo_env_to_vec_env_v1(self.env)
        vec_env = ss.concat_vec_envs_v1(
            vec_env,
            num_vec_envs=self.num_vec_envs,
            num_cpus=1,
            base_class="stable_baselines3"
        )
        return vec_env

    def train(self):
        """
        Initialiseer en train het PPO-model met geoptimaliseerde hyperparameters.
        """
        print("Start training...")
        self.model = PPO(
            CnnPolicy,
            self.vec_env,
            verbose=self.verbose,
            batch_size=self.batch_size,
            n_steps=128,
            n_epochs=4,
            learning_rate=2.5e-4,
            ent_coef=0.01,
            gae_lambda=0.95,
            gamma=0.99,
            clip_range=0.1,
            vf_coef=0.5,
            max_grad_norm=0.5,
            tensorboard_log="./ppo_warlords_tensorboard/"
        )
        self.model.learn(total_timesteps=self.total_timesteps)
        print("Training gereed.")

    def save(self, model_name=None):
        """
        Sla het getrainde model op, optioneel met custom naam.
        """
        if self.model is None:
            raise ValueError("Model is nog niet getraind.")
        if model_name is None:
            timestamp = time.strftime("%Y%m%d-%H%M%S")
            model_name = f"warlords_ppo_model_{timestamp}"
        self.model.save(model_name)
        print(f"Model opgeslagen als {model_name}")

    def load(self, path):
        """
        Laad een eerder getraind PPO-model.
        """
        self.model = PPO.load(path)
        print(f"Model geladen van {path}")


In [ ]:
agent2 = MARLAgentPPO()
agent2.train()
agent2.save("ppo_model_warlords_4m")
save_path = "/content/drive/MyDrive/MARL_models"
import os

# Zorg dat de map bestaat
os.makedirs(save_path, exist_ok=True)

# Sla model op met een custom naam en pad
model_name = os.path.join(save_path, "warlords_ppo_model_4m")
agent2.save(model_name)

Start training...
Using cuda device
Wrapping the env in a VecTransposeImage.


/usr/local/lib/python3.11/dist-packages/stable_baselines3/common/vec_env/base_vec_env.py:78: UserWarning: The `render_mode` attribute is not defined in your environment. It will be set to None.
  warnings.warn("The `render_mode` attribute is not defined in your environment. It will be set to None.")


Streaming output truncated to the last 5000 lines.
|    approx_kl            | 0.0024425047 |
|    clip_fraction        | 0.0933       |
|    clip_range           | 0.1          |
|    entropy_loss         | -1.72        |
|    explained_variance   | 0.219        |
|    learning_rate        | 0.00025      |
|    loss                 | -0.0296      |
|    n_updates            | 14516        |
|    policy_gradient_loss | -0.00833     |
|    value_loss           | 1.77e-05     |
------------------------------------------
------------------------------------------
| time/                   |              |
|    fps                  | 591          |
|    iterations           | 3631         |
|    time_elapsed         | 6289         |
|    total_timesteps      | 3718144      |
| train/                  |              |
|    approx_kl            | 0.0040649907 |
|    clip_fraction        | 0.151        |
|    clip_range           | 0.1          |
|    entropy_loss         | -1.72        |
|  

<div style="background-color:white; color:black; padding: 10px;">
    <strong>(Julia) PPO Agent (WEET NIET OF DEZE OOK OPRECHT WERKT MET TRAINING)</strong>
</div>

In [ ]:
# Model saven
save_path = "/content/drive/MyDrive/MARL_models"
import os

# Zorg dat de map bestaat
os.makedirs(save_path, exist_ok=True)

# Sla model op met een custom naam en pad
model_name = os.path.join(save_path, "warlords_ppo_model")
agent.save(model_name)


Model opgeslagen als /content/drive/MyDrive/MARL_models/warlords_ppo_model


In [ ]:
agent.load("ppo_model_warlords")

Model geladen van ppo_model_warlords


In [ ]:
model_path = "/content/drive/MyDrive/MARL_models/warlords_ppo_model.zip"
agent.load(model_path)

Model geladen van /content/drive/MyDrive/MARL_models/warlords_ppo_model.zip



---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H5: Training en Hyperparameter Search</strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;5.1: Training</h3>

<div style="background-color:white; color:black; padding: 10px;">
    <strong>(Rogier) Agent 1: PPO - Proximal Policy Optimization</strong></div>

### Hoe verloopt de training van een PPO-agent?

De training van een PPO-agent (Proximal Policy Optimization) in een omgeving zoals Atari Warlords bestaat uit een reeks vaste stappen die gericht zijn op stabiliteit en efficiënt leren. Tijdens elke trainingscyclus verzamelt de agent zogenaamde 'rollouts': sequenties van observaties, acties en rewards die worden gegenereerd door het huidige beleid. Op basis van deze trajecten berekent PPO met Generalized Advantage Estimation (GAE) hoe gunstig elke actie was ten opzichte van de verwachting, wat helpt om het leerproces stabieler en nauwkeuriger te maken.

De kern van PPO ligt in het voorzichtig updaten van het beleid. In plaats van grote stappen (die tot instabiliteit kunnen leiden), worden policy-updates beperkt door een clipping-mechanisme. Dit voorkomt dat het geleerde beleid te veel verandert per trainingsstap, waardoor de kans op catastrofaal “vergeten” sterk wordt verminderd. De verzamelde rollouts worden verdeeld over mini-batches en gedurende meerdere epochs gebruikt om zowel het beleid (policy) als de value-functie (critic) te verbeteren. Tegelijkertijd stimuleert PPO via een entropiebonus dat de agent in het begin veel blijft exploreren, wat de kans op het vinden van sterke strategieën vergroot.

Tijdens en na de training wordt de agent regelmatig geëvalueerd zonder extra ruis, zodat duidelijk wordt hoe effectief het geleerde beleid is in de praktijk. Door deze aanpak is PPO bijzonder geschikt voor complexe, visuele en multi-agent omgevingen, waarbij gecontroleerd leren en robuuste prestaties essentieel zijn.


<div style="background-color:white; color:black; padding: 10px;">
    <strong>Omgeving opzetten</strong>
</div>


In deze stap initialiseren we de multi-agent Warlords-omgeving met behulp van PettingZoo. We passen verschillende bewerkingen toe op de raw game frames:
- `warlords_v3.parallel_env()`: Laadt de 4-speler Warlords-omgeving waarbij alle agents gelijktijdig acties nemen.
- `ss.black_death_v3(env)`: Zorgt ervoor dat agents die "dood" gaan toch in de omgeving blijven als placeholder, zodat de agent-count altijd gelijk blijft.
- `ss.color_reduction_v0(env, mode='full')`: Zet de gekleurde frames om naar grijswaarden, waardoor de inputdimensie kleiner wordt en het leren efficiënter.
- `ss.resize_v1(env, x_size=84, y_size=84)`: Schaal de beelden terug naar 84x84 pixels (standaard in Atari-RL-onderzoek).
- `ss.frame_stack_v1(env, 4)`: Stapelt de laatste 4 frames, zodat het model ook bewegingsinformatie kan oppikken (belangrijk bij visuele input).

---

In [ ]:
# 1. Maak de parallelle multi-agent omgeving
env = warlords_v3.parallel_env()            # 4-player Warlords omgeving
env = ss.black_death_v3(env)               # Houd agents in env, zelfs na "dood"
env = ss.color_reduction_v0(env, mode='full')  # Converteer naar grijswaarden:contentReference[oaicite:22]{index=22}
env = ss.resize_v1(env, x_size=84, y_size=84)   # Resize frames naar 84x84 pixels
env = ss.frame_stack_v1(env, 4)                # Stack 4 opvolgende frames:contentReference[oaicite:23]{index=23}

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Converteren naar een vectorized env voor Stable-Baselines3</strong>
</div>

Stable-Baselines3 werkt het best met vectorized environments, waarbij meerdere instanties van de omgeving parallel kunnen draaien:
- `ss.pettingzoo_env_to_vec_env_v1(env)`: Zet de PettingZoo-omgeving om naar een formaat dat door SB3 wordt herkend.
- `ss.concat_vec_envs_v1(...)`: Combineert meerdere vectorized omgevingen tot één, zodat batchgewijze training mogelijk is (hier draaien 2 parallelle omgevingen op 1 CPU).

---

In [ ]:
# 2. Converteer naar een vectorized env voor Stable-Baselines3
vec_env = ss.pettingzoo_env_to_vec_env_v1(env)
vec_env = ss.concat_vec_envs_v1(vec_env, num_vec_envs=2, num_cpus=1, base_class="stable_baselines3")

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Defineren en trainen van het PPO model</strong>
</div>

We definiëren het PPO-model dat gebruikmaakt van een convolutioneel neuraal netwerk (CnnPolicy), geschikt voor visuele input zoals Atari-frames:
- `PPO(CnnPolicy, vec_env, ...)`: Initialiseer het PPO-algoritme met de eerder gemaakte vectorized environment.
- `total_timesteps`: Bepaalt hoe lang het model traint. Meer timesteps betekent meestal beter getrainde agenten.
- `model.learn(...)`: Start het daadwerkelijke leerproces. Het model verzamelt data, leert van ervaringen en past het beleid continu aan.

---

In [ ]:
# 3. Definieer en train het PPO model
model = PPO(CnnPolicy, vec_env, verbose=3, batch_size=256)  # CNN-beleid voor visuele input:contentReference[oaicite:24]{index=24}
total_timesteps = 10_000_000   # kies het aantal trainingstimesteps
print("Start training...")
model.learn(total_timesteps=total_timesteps)
print("Training gereed.")

Using cuda device
Wrapping the env in a VecTransposeImage.


/usr/local/lib/python3.11/dist-packages/stable_baselines3/common/vec_env/base_vec_env.py:78: UserWarning: The `render_mode` attribute is not defined in your environment. It will be set to None.
  warnings.warn("The `render_mode` attribute is not defined in your environment. It will be set to None.")


Start training...
------------------------------
| time/              |       |
|    fps             | 685   |
|    iterations      | 1     |
|    time_elapsed    | 23    |
|    total_timesteps | 16384 |
------------------------------
-----------------------------------------
| time/                   |             |
|    fps                  | 557         |
|    iterations           | 2           |
|    time_elapsed         | 58          |
|    total_timesteps      | 32768       |
| train/                  |             |
|    approx_kl            | 0.007993929 |
|    clip_fraction        | 0.0178      |
|    clip_range           | 0.2         |
|    entropy_loss         | -1.79       |
|    explained_variance   | -1.78       |
|    learning_rate        | 0.0003      |
|    loss                 | 0.00604     |
|    n_updates            | 10          |
|    policy_gradient_loss | -0.00251    |
|    value_loss           | 0.00127     |
-----------------------------------------
---------

KeyboardInterrupt: 

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Opslaan van het getrainde model met timestamp</strong>
</div>

Na de training slaan we het model op, waarbij automatisch een timestamp aan de bestandsnaam wordt toegevoegd:
- `time.strftime(...)`: Maakt een string van de huidige datum en tijd.
- `model.save(model_name)`: Slaat het getrainde model op onder een unieke naam, zodat je het later gemakkelijk kunt laden en evalueren.

---

In [ ]:
# 4. Sla het getrainde model op
timestamp = time.strftime("%Y%m%d-%H%M%S")
model_name = f"warlords_ppo_model_{timestamp}"
model.save(model_name)
print(f"Model opgeslagen als {model_name}")

In [ ]:
import time
import supersuit as ss
from stable_baselines3 import PPO
from stable_baselines3.ppo import CnnPolicy
from pettingzoo.atari import warlords_v3

# 1. Maak de parallelle multi-agent omgeving
env = warlords_v3.parallel_env()            # 4-player Warlords omgeving
env = ss.black_death_v3(env)               # Houd agents in env, zelfs na "dood"
env = ss.color_reduction_v0(env, mode='full')  # Converteer naar grijswaarden:contentReference[oaicite:22]{index=22}
env = ss.resize_v1(env, x_size=84, y_size=84)   # Resize frames naar 84x84 pixels
env = ss.frame_stack_v1(env, 4)                # Stack 4 opvolgende frames:contentReference[oaicite:23]{index=23}

# 2. Converteer naar een vectorized env voor Stable-Baselines3
vec_env = ss.pettingzoo_env_to_vec_env_v1(env)
vec_env = ss.concat_vec_envs_v1(vec_env, num_vec_envs=2, num_cpus=1, base_class="stable_baselines3")

# 3. Definieer en train het PPO model
model = PPO(CnnPolicy, vec_env, verbose=3, batch_size=256)  # CNN-beleid voor visuele input:contentReference[oaicite:24]{index=24}
total_timesteps = 2_000_000   # kies het aantal trainingstimesteps
print("Start training...")
model.learn(total_timesteps=total_timesteps)
print("Training gereed.")

# 4. Sla het getrainde model op
timestamp = time.strftime("%Y%m%d-%H%M%S")
model_name = f"warlords_ppo_model_{timestamp}"
model.save(model_name)
print(f"Model opgeslagen als {model_name}")



<div style="background-color:white; color:black; padding: 10px;">
    <strong>Toelichting op de output</strong>
</div>

## Interpretatie van de PPO Trainingsoutput

Tijdens het trainen van het PPO-model worden verschillende statistieken gelogd. Hieronder lichten we de belangrijkste waarden uit, zodat duidelijk wordt wat ze betekenen en hoe ze geïnterpreteerd kunnen worden:

| **Metric**               | **Waarde**      | **Toelichting**                                                                                                                                             |
|------------------------- |-----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **fps**                  | 472             | *Frames per second* – het aantal frames (omgevingsstappen) dat per seconde wordt verwerkt. Een hoge waarde betekent dat het trainen efficiënt verloopt.     |
| **iterations**           | 8               | Aantal PPO-updates die zijn uitgevoerd sinds de start van de training.                                                                                      |
| **time_elapsed**         | 277             | Totale verstreken tijd (in seconden) sinds het begin van de training.                                                                                       |
| **total_timesteps**      | 131072          | Het aantal omgevingsstappen (frames/acties) dat het model tot nu toe heeft gezien.                                                                          |

### Train-metrics

| **Metric**                   | **Waarde**        | **Toelichting**                                                                                                                                             |
|------------------------------|-------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **approx_kl**                | 0.00897           | Gemiddelde KL-divergence tussen het oude en het nieuwe beleid. Een lage waarde duidt op kleine wijzigingen per update, wat zorgt voor stabiel leren.         |
| **clip_fraction**            | 0.088             | Percentage van de policy-updates die beperkt (“geclipped”) werden. Veel clipping betekent dat het model te grote stappen probeert te maken (mogelijk instabiel). |
| **clip_range**               | 0.2               | Maximale toegestane relatieve verandering per update (standaard bij PPO).                                                                                   |
| **entropy_loss**             | -1.77             | De entropie van het beleid: een maat voor exploratie. Een lagere (meer negatieve) waarde betekent dat het beleid voorspelbaarder wordt (minder exploratie).   |
| **explained_variance**       | -0.0209           | Meet hoe goed de value-functie de daadwerkelijke rewards voorspelt (1.0 = perfect, 0 = slecht, <0 = erger dan gokken).                                      |
| **learning_rate**            | 0.0003            | De snelheid waarmee het model leert; hogere waardes versnellen leren, maar kunnen leiden tot instabiliteit.                                                 |
| **loss**                     | -0.0226           | Totale loss-functie (mix van policy, value en entropy loss) die wordt geminimaliseerd tijdens training.                                                     |
| **n_updates**                | 70                | Totaal aantal policy-updates tot nu toe uitgevoerd.                                                                                                         |
| **policy_gradient_loss**     | -0.0133           | De bijdrage van de policy-gradient aan de totale loss. Negatiever betekent sterkere updates in de richting van meer reward.                                 |
| **value_loss**               | 0.00265           | Fout van de value-functie; hoe kleiner, hoe beter het model toekomstige rewards kan voorspellen.                                                            |

---

### Richting van de Waarden: Wat wil je zien?

- **fps**: Hoger is beter – snellere training.
- **iterations / n_updates / total_timesteps**: Hoger betekent meer getraind (maar let op overfitting).
- **approx_kl**: Laag (bijv. tussen 0.01 en 0.05) = stabiel leren. Te hoog: beleid verandert te snel en wordt instabiel; te laag: leren gaat traag.
- **clip_fraction**: Typisch rond 0.1–0.3. Te hoog kan instabiliteit betekenen, te laag kan duiden op te weinig leereffect.
- **entropy_loss**: Minder negatief betekent meer exploratie. Naarmate het model zekerder wordt, daalt de entropie. Een te lage entropie betekent mogelijk te weinig exploratie.
- **explained_variance**: Hoger is beter. Richtwaarde: richting 1.0 is perfect, 0 is slecht, <0 betekent dat de value-functie slechter presteert dan gokken.
- **learning_rate**: Hogere waardes versnellen het leren maar verhogen het risico op instabiliteit.
- **loss / value_loss**: Lager is beter – betekent dat het model beter de returns/value weet te voorspellen en de policy verbetert.

#### **Samenvatting**
Deze statistieken geven inzicht in hoe snel en stabiel het PPO-model leert. Let vooral op de **KL-divergence** (voor stabiliteit), **clip_fraction** (voor learning dynamics), en **explained_variance** (voor de nauwkeurigheid van de value-voorspelling).


<a name='1.1'></a>
<h3>&sect;5.2: Selectie en tuning van hyperparameters</h3>

a.	Experimenteer met hyperparameters (zoals bijvoorbeeld learning rates of exploration-exploitation afwegingen). Documenteer dit ook.

##### Experimenteren met hyperparameters

### **Hyperparameters**

##### ***Overzicht keuzes***





| Hyperparameter        | Waarde    | Uitleg                                                                                 |
|----------------------|-----------|----------------------------------------------------------------------------------------|
| `Policy`              | CnnPolicy | Gebruikt convolutieneurale netwerken, geschikt voor beeldinput zoals bij Atari-games   |
| `learning_rate`        | 2.5e-4    | Standaardwaarde voor PPO in Atari-omgevingen; zorgt voor een goede balans tussen snel leren en stabiliteit |
| `gamma`                | 0.99      | Kortingfactor; waardeert toekomstige beloningen bijna net zo hoog als directe beloning |
| `n_steps`              | 128       | Aantal stappen per update; bepaalt hoeveel ervaringen per batch worden verzameld       |
| `batch_size`           | 256       | Grootte van elke batch die wordt gebruikt tijdens het updaten van het beleid           |
| `ent_coef (entropie)`  | 0.01      | Coëfficiënt voor exploratie; hogere waarde zorgt voor meer exploratiegedrag            |
| `verbose`              | 1         | Zorgt voor gedetailleerde logging tijdens het trainen                                  |

#### **Experimenteren met hyperparameters**
Tijdens het trainen heb ik geëxperimenteerd met verschillende waarden voor de learning rate, entropie-coëfficiënt en batch size:
- **Learning rate:** Een hogere learning rate zorgde voor snellere training, maar maakte het model soms instabiel. 2.5e-4 bleek een goed compromis.
- **Entropie-coëfficiënt (ent_coef):** Met een hogere waarde ging de agent meer experimenteren, maar duurde het langer voordat hij goed leerde. 0.01 gaf een goede balans.
- **Batch size:** Grotere batches zorgden voor stabielere updates, maar vroegen meer geheugen.

De uiteindelijke hyperparameterkeuzes zijn gebaseerd op wat het beste werkte voor deze specifieke omgeving en op basis van literatuur.




---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H6: Evaluatie en Vergelijking</strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;6.1: Evaluatie t.o.v. baseline</h3>

b.	Evalueer de prestaties van je model en vergelijk deze – indien mogelijk – met een baseline. Welke voordelen biedt RL in dit specifieke scenario?

<a name='1.1'></a>
<h3>&sect;6.2: Analyse met metrics</h3>

c.	Visualiseer de resultaten met grafieken en andere tools. Denk bijvoorbeeld aan reward-curves en stabiliteit van training.

<div style="background-color:white; color:black; padding: 10px;">
    <strong>Metrics Model met 200_000 time stamps</strong>
</div>

In [ ]:
# drive 1000000 time step model

model_path = "/content/drive/MyDrive/MARL_models/warlords_ppo_model_4m.zip"
model_name = model_path

In [ ]:

import numpy as np

# 1. Laad het opgeslagen model
model = PPO.load(model_name) # Tijdelijk, hier getrainde model zetten huts

# 2. Maak een nieuwe omgeving voor evaluatie (parallel API)
eval_env = warlords_v3.parallel_env(render_mode=None)  # geen render, alleen data
eval_env = ss.black_death_v3(eval_env)
eval_env = ss.color_reduction_v0(eval_env, mode='full')
eval_env = ss.resize_v1(eval_env, x_size=84, y_size=84)
eval_env = ss.frame_stack_v1(eval_env, 4)

# 3. Functie om een episode te spelen met 1 RL agent vs 3 random agents
def play_one_episode(env, agent_policy, render=False):
    obs, info = env.reset(seed=None)  # <-- Let op: 2 outputs!
    episode_rewards = {agent: 0 for agent in env.possible_agents}
    step = 0
    while True:
        actions = {}
        for agent, agent_obs in obs.items():
            if agent == env.possible_agents[0]:
                action, _state = agent_policy.predict(agent_obs, deterministic=True)
            else:
                action = env.action_space(agent).sample()
            actions[agent] = action
        obs, rewards, terminations, truncations, infos = env.step(actions)  # <-- 5 outputs
        if render:
            env.render()
        for agent, r in rewards.items():
            episode_rewards[agent] += r
        if all(terminations.values()) or any(truncations.values()):
            break
        step += 1
    return episode_rewards

# 4. Speel meerdere episodes en verzamel resultaten
num_test_episodes = 20
total_rewards = {agent: 0 for agent in eval_env.possible_agents}
wins = {agent: 0 for agent in eval_env.possible_agents}

for ep in range(1, num_test_episodes+1):
    ep_rewards = play_one_episode(eval_env, model, render=False)
    print(f"Episode {ep} rewards: {ep_rewards}")
    for ag, r in ep_rewards.items():
        total_rewards[ag] += r
    for ag, r in ep_rewards.items():
        if r > 0:  # +1 winnaar
            wins[ag] += 1

# 5. Bereken gemiddelde reward per agent over alle test-episodes
avg_rewards = {agent: total_rewards[agent]/num_test_episodes for agent in total_rewards}
print("\nGemiddelde reward per agent over", num_test_episodes, "episodes:")
for ag, r in avg_rewards.items():
    print(f"  {ag}: {r:.2f}")
print("Aantal keer gewonnen (laatste overlevende) per agent:", wins)


Episode 1 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(1.0), 'third_0': np.float64(-1.0), 'fourth_0': np.float64(-1.0)}
Episode 2 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(-1.0), 'third_0': np.float64(1.0), 'fourth_0': np.float64(-1.0)}
Episode 3 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(1.0), 'third_0': np.float64(-1.0), 'fourth_0': np.float64(-1.0)}
Episode 4 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(-1.0), 'third_0': np.float64(-1.0), 'fourth_0': np.float64(1.0)}
Episode 5 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(-1.0), 'third_0': np.float64(-1.0), 'fourth_0': np.float64(1.0)}
Episode 6 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(-1.0), 'third_0': np.float64(-1.0), 'fourth_0': np.float64(1.0)}
Episode 7 rewards: {'first_0': np.float64(-1.0), 'second_0': np.float64(-1.0), 'third_0': np.float64(1.0), 'fourth_0': np.float64(-1.0)}
Episode 8 rewards: {'first_0': np.float64

In [ ]:
# Hier kijken hoe ik het in een gifje kan krijgen te zien

<a name='1.1'></a>
<h3>&sect;6.3: Visualisatie van resultaten</h3>

c.	Visualiseer de resultaten met grafieken en andere tools. Denk bijvoorbeeld aan reward-curves en stabiliteit van training.


---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H7: Rapportage en Reflectie</strong></h2>
</div>

<a name='1.1'></a>
<h3>&sect;7.1: Methodologie en aanpak</h3>

<a name='1.1'></a>
<h3>&sect;7.2: Samenvatting van resultaten</h3>

<a name='1.1'></a>
<h3>&sect;7.3: Reflectie op model, prestaties en uitbreidingsmogelijkheden</h3>


a.	Schrijf een rapport waarin je je probleemstelling, aanpak, resultaten en conclusies presenteert. Reflecteer op mogelijke uitbreidingen.


---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H8: Literatuurlijst</strong></h2>
</div>

- Scharwächter, V. (2024, 15 juli). *Probleemanalyse maken voor je scriptie | Betekenis & Voorbeeld.* Scribbr. https://www.scribbr.nl/starten-met-je-scriptie/probleemanalyse/
- PPO — Stable Baselines3 2.7.0a0 documentation. (z.d.). https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html


---
<div style="background-color:LightBlue; text-align:center; vertical-align:middle; padding:50px 0; margin-top:5px; margin-bottom:5px">
    <h2 id="eda-title" style="color:black; font-family: Verdana, sans-serif; font-size: 25px;"><strong>H9: Beoordelingscriteria</strong></h2>
</div>

Je werk wordt beoordeeld op de volgende aspecten:
1.	**Keuze van algoritme en trainingsstrategie**: Is het gekozen algoritme geschikt voor de omgeving? Is de trainingsmethode van de agent relevant voor deze setting? Wordt er rekening gehouden met de aanwezigheid van meerdere agenten?
2.	**Technische diepgang**: Worden neural network-componenten (indien van toepassing) correct toegepast? Is er aandacht voor alle cruciale onderdelen van de DRL-pijplijn (zoals experience replay en policy updates)?
3.	**Implementatie en testen**: Is het multi-agent Reinforcement Learning-algoritme correct geïmplementeerd, met goed gestructureerde en werkende code? Is het getest in een multi-agentomgeving?
4.	**Rapportage**: Zijn de methodologie en keuzes goed onderbouwd met wetenschappelijke literatuur? Is het rapport helder en gestructureerd?
5.	**Reproduceerbaarheid**: Is de code duidelijk, goed gedocumenteerd en eenvoudig te reproduceren?
6.	**Bonuspunten**: Aan het einde van het project neemt je agent het in een toernooi op tegen je klasgenoten. (De details hierover volgen later.) Het winnende team van het toernooi krijgt 5 bonuspunten bovenop het aantal behaalde punten met de opdracht.
